File: Dynamic_Array/javascript/dynamicArray.js
export default class DynamicArray {
  size = 0;
  capacity = 10;
  array;
  constructor() {
    this.array = new Array(this.capacity);
  }

  add(data) {
    if (this.size >= this.capacity) {
      this.grow();
    }
    this.array[this.size] = data;
    this.size++;
    if (this.size >= this.capacity) {
      this.grow();
    }
  }

  insert(index, data) {
    if (this.size >= this.capacity) {
      this.grow();
    }

    for (let i = this.size; i > index; i--) {
      this.array[i] = this.array[i - 1];
    }
    this.array[index] = data;
    this.size++;
  }

  delete(data) {
    for (let i = 0; i < this.size; i++) {
      if (this.array[i] == data) {
        for (let n = i; n < this.size; n++) {
          this.array[n] = this.array[n + 1];
        }
        this.size--;
        if (this.size < parseInt(this.capacity / 3)) {
          this.shrink();
        }
        break;
      }
    }
  }

  search(data) {
    let index;
    for (let i = 0; i < this.size; i++) {
      if (this.array[i] == data) {
        index = i;
        break;
      }
    }
    return index;
  }

  grow() {
    let nCapacity = 10 * 2;
    let nArray = new Array(nCapacity);
    for (let i = 0; i < this.size; i++) {
      nArray[i] = this.array[i];
    }
    this.array = nArray;
    this.capacity = nCapacity;
  }

  shrink() {
    let nCapacity = this.capacity / 2;
    let nArray = new Array(nCapacity);
    for (let i = 0; i < this.size; i++) {
      nArray[i] = this.array[i];
    }
    this.array = nArray;
    this.capacity = nCapacity;
  }

  isEmpty() {
    return this.size == 0;
  }
}

File: Dynamic_Array/c++/dynamicArray.cpp
#include <iostream>
using namespace std;

class DynamicArray
{
private:
    string toString(){
        string arr;
        for (size_t i = 0; i < size; i++)
        {
            arr += array[i]+",";
        }
        arr.pop_back();
        arr = "["+arr+"]";
        return arr;     
    }

    void grow(){
        int ncapacity = capacity*2;
        string* narray = new string[ncapacity];
        for(int i=0;i<size;i++){
            narray[i]= array[i];
        }
        array = narray;
        capacity = ncapacity;
    }

    void shrink(){
        int ncapacity = int(capacity/2);
        string* narray = new string[ncapacity];
        for(int i=0;i<size;i++){
            narray[i]= array[i];
        }
        array = narray;
        capacity = ncapacity;
    }
public:
    int size=0;
    int capacity = 10;
    string* array;

    DynamicArray(){
        array = new string[capacity];
    };
    
    void displayData(){
        cout<<"SIZE: "<<size<<endl;
        cout<<"CAP: "<<capacity<<endl;
        cout<<"Array: "<<toString()<<endl;
    }

    void add(string data){
        array[size] = data;
        size++;
        if(size>=capacity){
            grow();
        }
    }

    void insert(int index, string data){
        for (size_t i = 4; i > 3; i--)
        {
            array[i] = array[i-1];
            // array[i] = array[i+1];
        }
        array[index] = data;
        size++;
        if(size>=capacity){
            grow();
        }
    }

    void remove(string data){
        for(int i=0;i<size;i++){
            if(array[i]==data){
                for(int n=i; n<size;n++){
                    array[n] = array[n+1];
                }
                size--;
                if(size<=int(capacity/3)){
                    shrink();
                }
                break;
            }
        }
        
    }

    int search(string data){
        for(int i=0;i<size;i++){
            if(array[i]==data){
                cout<<"Index for "<< data <<" is: "<< i<<endl;
            }
        }
        return -1;
    }
};
